### ISSUE TITLE
Performance Issues

### ISSUE THREAD

**Issue by tkell** [algorithmic-music-exploration/amen#55]
...is my computer amazingly slow?  How long should it take for librosa to analyze a five minute long wav file?

I am getting like 45 second to a minute to create an Audio object, and even longer for my apparently awful synthesis code to run.  Has anyone had comparable experiences?

**Comment by tkell** [comment]
a)  My machine takes a very long time to make an Audio object out of a five minute track.  Unsure where the problem is

b)  I'm re-generating the zero crossings _every time_ in synthesize.py.  I am going to fix this

c) We should have picked a better test file.  I may replace the pure-but-short amen.wav with some 3 minute jungle tune.

**Comment by tkell** [comment]
Took a crack at b) with #56

**Comment by bmcfee** [comment]
I can imagine a few bottlenecks, but yeah, zc calculation could add up.  Caching makes sense.

Have you profiled it?

**Comment by tkell** [comment]
@bmcfee, I did some very low-rent profiling.  For a 5 minute track, zc calculations takes just under a second, and I was doing it for every beat.

I'm actually going to rip out the above code and just calculate zero indexes whenever we make an Audio object.  It clears up a ton of logic, and I think it is safe to assume that if we're loading a track, we're going to want to write it out.

**Comment by tkell** [comment]
@bmcfee, I've updated #56 and it is so much better now.  Will do more profiling tomorrow

**Comment by bmcfee** [comment]
Cool, makes sense.

**Comment by tkell** [comment]
An example run of the profiling for loading our two-bar test track:
```
librosa load 0.0368201732635
filepath, sample rate, duration, etc 0.00219583511353
resampling ** 2.06249713898**
zero crossings 0.0202448368073
all features 0.615534067154
timings 0.191967010498
Librosa load 3.09944152832e-06
```

Basically we're getting owned on resampling.  @bmcfee, will your new resampler help with that at all?  And what would the cost be with not resampling and using the track's SR to do analysis?

**Comment by bmcfee** [comment]
> Basically we're getting owned on resampling.

Do you know which resampler it's using?  (That is, do you get the scikits.samplerate warning on load?)

And what rates is it converting from/to?

How many resampling ops is it doing?  2sec is a lot for 2 bars of audio, which makes me think that it's either using scipy.signal.resample (slow), or doing way too many resample ops.

> will your new resampler help with that at all?

Yeah, probably.  If the above numbers are in fact using scikits.samplerate (and not scipy), then there might be about a 10% speed boost, which isn't much.  If it's using scipy, then yes, it will definitely be better using resampy instead.

If you're feeling intrepid, you can check out [this PR on librosa](https://github.com/bmcfee/librosa/pull/324), and do
```bash
pip install --pre resampy
```
to see if the new resampler helps.  It should all work ; merge is waiting on a final stable release of resampy 0.1.0, so any additional testing would be greatly appreciated.

**Comment by bmcfee** [comment]
Sorry, forgot to hit the last point:

> And what would the cost be with not resampling and using the track's SR to do analysis?

That'd be a big hornet's nest.  Basically, all of the timing parameters depend on the sampling rate, and it would be a real pain to thread that calculation through every stage of the analysis.  For some ops (cqt, chroma), there's no avoiding resampling -- it's part of the analysis itself.

But analysis is a one-time cost, most of the features are linear time (maybe quadratic for segmentation, but only after subsampling by beat synchronization), and the bulk of that code is pretty well optimized at this point.

**Comment by tkell** [comment]
Rad.  As per our twitter convo, I will try out `resampy` this weeked / next week, and wrap the creation of an Audio object in https://pypi.python.org/pypi/tqdm.

**Comment by tkell** [comment]
Closed #56 to fix b).

**Comment by tkell** [comment]
@bmcfee, found an issue that I don't understand with resampy, but it mostly works!

If you create an Audio straight from raw samples, with very small data, like so:

```
faux_samples = np.array([[1.0, -1.0, 0.0, 1.0, 0.0, -1.0, 1.0],
                          [1.0, -1.0, 0.0, 1.0, 0.0, -1.0, 1.0]])
faux_audio = Audio(raw_samples=faux_samples)
```

You get yelled at when calculating the chroma:

```
Traceback (most recent call last):
  File "/home/thor/Code/amen-env/local/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/thor/Code/amen-env/local/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/thor/Code/amen-env/local/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/thor/Code/amen/tests/test_timing.py", line 33, in <module>
    faux_audio = Audio(raw_samples=faux_samples)
  File "/home/thor/Code/amen/amen/audio.py", line 82, in __init__
    self.features = self._create_features()
  File "/home/thor/Code/amen/amen/audio.py", line 154, in _create_features
    features['chroma'] = self._get_chroma()
  File "/home/thor/Code/amen/amen/audio.py", line 209, in _get_chroma
    chroma_cq = librosa.feature.chroma_cqt(mono_samples)
  File "/home/thor/Code/librosa/librosa/feature/spectral.py", line 933, in chroma_cqt
    real=False))
  File "/home/thor/Code/librosa/librosa/core/constantq.py", line 252, in cqt
    scale=True)
  File "/home/thor/Code/librosa/librosa/core/audio.py", line 274, in resample
    y_hat = resampy.resample(y, orig_sr, target_sr, filter=res_type, axis=-1)
  File "/home/thor/Code/amen-env/local/lib/python2.7/site-packages/resampy/core.py", line 104, in resample
    y_2d = y.swapaxes(0, axis).reshape((y.shape[axis], -1))
ValueError: total size of new array must be unchanged
```

This _appears_ to be related to the size of the array - the very small fake array fails, whereas when I do:

```
EXAMPLE_FILE = example_audio_file()
stereo_audio = Audio(EXAMPLE_FILE)
faux_audio = Audio(raw_samples=stereo_audio.raw_samples)
```

I am pretty sure I can patch around this tomorrow, as far as the Amen tests are concerned.  Will update as I do it.

**Comment by bmcfee** [comment]
Oh, that's an interesting one.

The issue, I think, is that the signal is too short to do recursive down sampling. Probably the should be a length check for this in librosa. (I suspect resampy is doing the right thing.)

Can you strip this example down to not depend on Amen, and post it as an issue to librosa?

**Comment by tkell** [comment]
Can do!

Also, things are less bad with resampy

```
librosa load 0.0455520153046
filepath, sample rate, duration, etc 0.00095009803772
resampling 0.702908039093 # !! down from 2.0
features, zc, timings 0.820599079132
```

**Comment by bmcfee** [comment]
Ok, librosa and resampy are being updated to explicitly handle the short-signal case.

All this means here is that the exception you get will be more informative -- cqt-based analysis on a signal as short as the one you're using will still fail.

It's a little tricky to pin down how long an input needs to be, since it depends on the sampling rate and range of the CQT bins.  A conservative estimate would be `len(y) >= hop_length` -- you can actually go a bit lower than that if you know exactly how many downsampling ops it will take (at least `n_octaves-1`).

TLDR: use longer test signals.



> Also, things are less bad with resampy

Sweet!  If you really want to crank the speed up, and don't care about high frequencies, you can resample with `res_type='kaiser_fast'`.  That ought to shave off another factor of 2 from the resample time.

**Comment by tkell** [comment]
With `kasier_fast`:

```
librosa load 3.09944152832e-06
filepath, sample rate, duration, etc 0.00243401527405
resampling 0.302961826324 # down from 2.0 when we started this thread.
features, timings, zc 0.752917051315
```

**Comment by tkell** [comment]
This is in #59, which needs a new release of librosa that knows what `kaiser_fast` is, so we can pass it to Travis.
